File: src/DTOs/Item.cs

using System.Collections.Generic;
using System.Linq;

namespace ItemTTT.DTOs
{
	public class Item
	{
		public string	Code			{ get; set; }
		public string	Name			{ get; set; }
		public string	DescriptionEN	{ get; set; }
		public string	DescriptionFR	{ get; set; }
		public string	DescriptionNL	{ get; set; }
		public int?		Price			{ get; set; }
		public bool		Active			{ get; set; }

		public Dictionary<Languages,string>	DetailsUrls	{ get; set; }
		public ItemPicture[]				Pictures	{ get; set; }
		public Utils.Translation[]			Options		{ get; set; }

		internal int 	MainImageNumber	= 1;  // nb: Not transmitted to client
		public ItemPicture		FirstImage	{ get { return ((Pictures??new ItemPicture[]{}).Length == 0) ? ItemPicture.Empty : Pictures.FirstOrDefault(); } }
		public ItemPicture		MainImage	{ get { return ((Pictures??new ItemPicture[]{}).Length == 0) ? ItemPicture.Empty : Pictures.Where(v=>v.Number == MainImageNumber).SingleOrDefault(); } }

		public Item(Models.Item src=null)
		{
			if( src == null )
				return;

			Code			= src.Code;
			Name			= src.Name;
			DescriptionEN	= src.DescriptionEN;
			DescriptionFR	= src.DescriptionFR;
			DescriptionNL	= src.DescriptionNL;
			MainImageNumber	= src.MainImageNumber ?? 1;
			Price			= src.Price;
			Active			= src.Active;
		}

		internal void ToModel(LogHelper logHelper, Models.Item dst)
		{
			dst.Code = Services.ItemController.GetUrlCodeS( string.IsNullOrWhiteSpace(Code) ? Name : Code );
			logHelper.AddLogMessage( $"Item.ToModel: Rewrote code from '{Code}' to '{dst.Code}'" );
			if( string.IsNullOrWhiteSpace(dst.Code) )
				throw new Utils.TTTException( "The code must be specified" );

			dst.Name			= ( string.IsNullOrWhiteSpace(Name)				? "" : Name				).Trim();
			dst.DescriptionEN	= ( string.IsNullOrWhiteSpace(DescriptionEN)	? "" : DescriptionEN	).Trim();
			dst.DescriptionFR	= ( string.IsNullOrWhiteSpace(DescriptionFR)	? "" : DescriptionFR	).Trim();
			dst.DescriptionNL	= ( string.IsNullOrWhiteSpace(DescriptionNL)	? "" : DescriptionNL	).Trim();
			dst.Price			= ( Price ?? -1 ) <= 0 ? null : Price;
			dst.Active			= Active;
		}
	}
}

File: app/src/DTOs/Item.ts

import * as common from "../Views/common";
import { BaseAutoItem, DictObj } from "../Utils/BaseAutoItem";
export { DictObj, DictKO } from "../Utils/BaseAutoItem";
import { Translation } from "./Translation";
import { ItemPicture } from "./ItemPicture";

export interface Item extends DictObj
{
	code		: string;
	active		: boolean;
	features	: Translation[];
	pictures	: ItemPicture[];
}

export class ItemKO extends BaseAutoItem
{
	public readonly code			: KnockoutObservable<string>;
	public readonly name			: KnockoutObservable<string>;
	public readonly descriptionEN	: KnockoutObservable<string>;
	public readonly descriptionFR	: KnockoutObservable<string>;
	public readonly descriptionNL	: KnockoutObservable<string>;
	public readonly active			: KnockoutObservable<boolean>;
	public readonly price			: KnockoutObservable<number>;

	constructor($container:JQuery, src?:Item, fieldNames?:string[])
	{
		if( src == null )
			src = { code:null, active:false, features:[], pictures:[] };
		super( $container, src, fieldNames );
		const self = this;

		if( self.price != null )
			common.utils.ensureInteger({ observable:self.price, canBeZero:false, canBeNull:true, mustBePositive:true });
	}

	public /*override*/ toDictObj(dict?:DictObj) : Item
	{
		const rc = super.toDictObj( dict );
		return <Item>rc;
	}
}
